backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs

```diff
@@ -59,90 +55,123 @@ public async Task<UserDto> AddTagsAsync(List<TagDto> tags)
+    {
+        var userId = _authService.UserId;
+
+        var user = _context.Users.SingleOrDefault(u => u.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");
+        var lesson = _context.Lessons.SingleOrDefault(l => l.Id == lessonId) ?? throw new ArgumentException($"Failed to find the lesson with id {lessonId}");
+
+        user.Lessons.Add(lesson);
+
+        await _context.SaveChangesAsync();
+
+        void AfterMapAction(Lesson o, LessonDto dto) => dto.SubscribersCount = _context.Lessons
+            .Select(t => new { Id = t.Id, SbCount = t.Subscribers.Count })
+            .FirstOrDefault(l => l.Id == lessonId)!.SbCount;
+
+        return _mapper.Map<Lesson, LessonDto>(lesson, options => options.AfterMap(AfterMapAction));
+    }
+
+    public async Task<string> UploadProfilePhoto(IFormFile file)
+    {
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == _authService.UserId);
+
+        if (user == null)
+        {
+            throw new ArgumentNullException("This user not found");
+        }
+
+        var fileDto = new NewEasySpeakFileDto()
+        {
+            Stream = file.OpenReadStream(),
+            FileName = file.FileName
+        };
+
+        var uploadFileDto = await _fileService.AddFileAsync(fileDto);
+        var profilePhoto = await _context.EasySpeakFiles.FirstOrDefaultAsync(f => f.Id == uploadFileDto.Id);
+
+        if (profilePhoto == null || profilePhoto.Url == null)
+        {
+            throw new ArgumentNullException("This file not found");
+        }
+
+        user.ImageId = uploadFileDto.Id;
+        profilePhoto.UserId = user.Id;
+        await _context.SaveChangesAsync();
+
+        return profilePhoto.Url;
+    }
+
+    private async Task<string> GetProfileImageUrl(long? imageId)
+    {
+        var profileImage = await _context.EasySpeakFiles.FirstOrDefaultAsync(f => f.Id == imageId);
+        return profileImage?.Url ?? "";
+    }
+
+    public async Task<TagDto[]> GetUserTags()
+    {
+        var userId = _authService.UserId;
+
+        var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(u => u.Id == userId);
+        var tags = user?.Tags ?? throw new ArgumentException($"Failed to find tags for user with id {userId}");
+
+        return _mapper.Map<TagDto[]>(tags);
+    }
+
+    public async Task<UserDto> UpdateUser(UserDto userDto)
+    {
+        var userId = _authService.UserId;
+        var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(a => a.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");
+
+        _mapper.Map(userDto, user);
+        user.Tags.Clear();
+
+        if (userDto.Tags != null)
+        {
+            user.Tags = userDto.Tags.Join(
+                   _context.Tags,
+                   dtoTags => dtoTags.Name,
+                   dbTags => dbTags.Name,
+                   (dto, dbTags) => dbTags).ToList();
+
+        }
```

Comment:
use after map and ret rid of Clear

Reply:
without `Clear` it didn't work but I'll try)

Reply:
Removing and adding values is not a good way for update data

Reply:
fixed without clearing the values